logic/src/commonMain/kotlin/com/wire/kalium/logic/data/user/UserRepository.kt

```diff
@@ -139,78 +156,105 @@ internal class UserDataSource internal constructor(
     private suspend fun updateSelfUserProviderAccountInfo(userDTO: UserDTO): Either<StorageFailure, Unit> =
         sessionRepository.updateSsoIdAndScimInfo(userDTO.id.toModel(), idMapper.toSsoId(userDTO.ssoID), userDTO.managedByDTO)
 
+    override suspend fun getKnownUser(userId: UserId): Flow<OtherUser?> =
+        userDAO.getUserByQualifiedID(qualifiedID = userId.toDao())
+            .map { userEntity ->
+                userEntity?.let { publicUserMapper.fromDaoModelToPublicUser(userEntity) }
+            }.onEach { otherUser ->
+                processFederatedUserRefresh(userId, otherUser)
+            }
+
+    /**
+     * Only in case of federated users and if it's expired or not cached, we fetch and refresh the user info.
+     */
+    private suspend fun processFederatedUserRefresh(userId: UserId, otherUser: OtherUser?) {
+        if (otherUser != null && otherUser.userType.isFederated()
+            && federatedUsersExpirationCache[userId]?.let { DateTimeUtil.currentInstant() > it } != false
```

Comment:
I wonder if we should use `ConcurrentMap.compute()` and put fetching user inside for the action to be performed atomically, because now reading and updating this map is in two separate actions so in theory two different threads can get in between, right?

Reply:
Hey good point! I think in that case, it would make sense to change the implementation and have it cached only once 🤔  Because we don't have a `compute` method in `ConcurrentMap` only a `computeIfAbsent`, and that not serves this purpose, because I would not be able to expire it based on time. And as Mr. T requested, I will not implement a mechanism to keep this up to date.

So I think we have 3 options:
- Not worrying about this case, I'm not seeing a real bad effect in case it happens, please help me here to see the downside.
- Change the implementation and cached in memory only, could be a, so it will be refreshed only once when the app is opened, could be a `ConcurrentSet` instead.
- Not caching at all, and that I would prefer not to do it, could be painful for the performance for someone working with federated users.

What do you think ?